libavcodec/libaribcaption.c

```diff
@@ -0,0 +1,1141 @@
+    }
+    if (avpkt->pts == AV_NOPTS_VALUE)
+        ctx->pts = ARIBCC_PTS_NOPTS;
+    else
+        ctx->pts = av_rescale_q(avpkt->pts, ctx->time_base, (AVRational){1, 1000});
+
+    status = aribcc_decoder_decode(ctx->decoder, avpkt->data, avpkt->size,
+                                   ctx->pts, &ctx->caption);
+    if (status == ARIBCC_DECODE_STATUS_ERROR) {
+        av_log(ctx, AV_LOG_ERROR,
+               "aribcc_decoder_decode() returned with error.\n");
+        return AVERROR(EAGAIN);
+    }
+    if (status == ARIBCC_DECODE_STATUS_NO_CAPTION) {
+        ff_dlog(ctx, "No caption.\n");
+        return avpkt->size;
+    } else {
+        ff_dlog(ctx, "type=%02x, flags=%x, lang=%03x\n",
+                ctx->caption.type, ctx->caption.flags, ctx->caption.iso6392_language_code);
+        ff_dlog(ctx, "region count = %d, start=%d.%d, duration=%d.%d\n",
+                ctx->caption.region_count,
+                (int)(ctx->caption.pts / 1000), (int)(ctx->caption.pts % 1000),
+                (int)((ctx->caption.wait_duration == ARIBCC_DURATION_INDEFINITE) ?
+                      -1 : ctx->caption.wait_duration / 1000),
+                (int)((ctx->caption.wait_duration == ARIBCC_DURATION_INDEFINITE) ?
+                      0 : ctx->caption.wait_duration % 1000));
+    }
+
+    switch (ctx->subtitle_type) {
+    case SUBTITLE_TEXT:
+        status = aribcaption_trans_text_subtitle(ctx);
+        break;
+
+    case SUBTITLE_ASS:
+        status = aribcaption_trans_ass_subtitle(ctx);
+        break;
+
+    case SUBTITLE_BITMAP:
+        status = aribcaption_trans_bitmap_subtitle(ctx);
+        break;
+
+    case SUBTITLE_NONE:
+    default:
+        status = 0;
+    }
+
+    if (status < 0) {
+        av_log(ctx, AV_LOG_ERROR, "Failed to set Subtitle: %s\n",
+               av_err2str(status));
+        aribcc_caption_cleanup(&ctx->caption);
+        return status;
+    }
+    if (status > 0) {
+        *got_sub_ptr = 1;
+        if (ctx->avpkt->pts != AV_NOPTS_VALUE)
+            sub->pts = av_rescale_q(ctx->avpkt->pts,
+                                    ctx->time_base, AV_TIME_BASE_Q);
+        if (ctx->caption.wait_duration == ARIBCC_DURATION_INDEFINITE)
+            sub->end_display_time = UINT32_MAX;
+        else
+            sub->end_display_time = (uint32_t)ctx->caption.wait_duration;
+    }
+
+    aribcc_caption_cleanup(&ctx->caption);
+    return avpkt->size;
+}
+
+static void aribcaption_flush(AVCodecContext *avctx)
+{
+    ARIBCaptionContext *ctx = avctx->priv_data;
+
+    if (!(avctx->flags2 & AV_CODEC_FLAG2_RO_FLUSH_NOOP)) {
+        if (ctx->decoder)
+            aribcc_decoder_flush(ctx->decoder);
+        if (ctx->renderer)
+            aribcc_renderer_flush(ctx->renderer);
+        ctx->readorder = 0;
+    }
+}
+
+static int aribcaption_close(AVCodecContext *avctx)
+{
+    ARIBCaptionContext *ctx = avctx->priv_data;
+
+    aribcaption_flush(avctx);
+
+    if (ctx->clut)
+        av_freep(&ctx->clut);
```

Comment:
You don't need to `if` around this; `av_freep(&x)` where `x` is null is a no-op.